enhancements/support-cluster-HA-API.md

```diff
@@ -0,0 +1,134 @@
+---
+title: support-cluster-HA-API
+authors:
+  - "@varshaprasad96"
+reviewers:
+  - "@jmrodri"
+  - "@ecordell"
+  - "@estroz"
+approvers:
+  - "@jmrodri"
+  - "@ecordell"
+  - "@estroz"
+creation-date: 2021-02-01
+status: implementable
+---
+
+# Support for Cluster High-availability Mode API
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+The goal of this enhancement is to describe the support for [cluster high-availability mode API][enhancement_cluster_operators] introduced by OpenShift. The enhancement proposes to introduce necessary helper functions which will enable users to identify if the OpenShift cluster can support high availability deployment mode of their operator or not.
+
+## Motivation
+
+OpenShift intends to provide an API to expose the high-availability expectations to operators in a cluster. The API will accomplish this by embedding additional fields in the [informational resource][informational_resource]. Providing the necessary helper functions which utilize the OpenShift API to expose the intended cluster topology, will enable users to identify the mode of operator deployment which the cluster can support. Operator authors can leverage this functionality and configure their operators to react accordingly.
+
+### Goals:
+
+- List the helper functions that read the OpenShift informational resource and return appropriate values to expose the high availability mode that the cluster can support.
+- Discuss on the appropriate location of the helper functions.
+
+### Non-Goals:
+
+- Discuss how OpenShift determines if the cluster can support HA/non-HA mode.
+- Discuss the behavior of operators in HA/non-HA mode.
+
+#### Story 1
+
+As an Operator author, I would like to know if the cluster can support HA or non-HA modes, so that I can configure my operator's behavior accordingly.
+
+## Design Details
+
+As an operator author it is helpful to know the cluster topology so that one can modify the behavior of operands. The library where this functionality resides, will provide users with a `GetClusterConfig` function that returns a `ClusterConfig` struct.
+
+The `ClusterConfig` struct can be defined to be:
```

Comment:
What benefit do we get from defining a new type instead of using the existing Infrastructure type?

Reply:
@dhellmann how will the API defined in this EP https://github.com/openshift/enhancements/blob/master/enhancements/single-node/cluster-high-availability-mode-api.md be accessible? I'm wondering if there's a need for a "wrapper" in operator-lib or library-go at all.

Reply:
This is already implemented here:
https://github.com/openshift/api/blob/master/config/v1/0000_10_config-operator_01_infrastructure.crd.yaml#L182

and example of usage:
https://github.com/openshift/cluster-authentication-operator/pull/391/files#diff-b6f4bf160fd7e801eadfbfac60b84bd00fcad21a0979c2d730549f10db015645R108

Reply:
As Ravid says, the API is implemented as 2 new fields on the Infrastructure configuration resource. It sounds like the proposal is to provide a convenience function to read that resource and extract the 2 fields to return in a new struct. Some level of help for operators that run on non-OpenShift clusters might be useful, since it means we could provide some defaulting behavior for those platforms. That's not what this proposal says, though (see my other suggestions elsewhere on the review).

Reply:
@dhellmann after investigating more I've come to the consensus that we probably don't need a convenience library at the moment. The existing API as pointed out by @ravidbro above is quite easy to access via normal k8s constructs and controller-runtime.

Simple controller-runtime example (just the act of getting it).
```golang
	// Simple query
	nn := types.NamespacedName{
		Name: "cluster",
	}
	infraConfig := &configv1.Infrastructure{}
	err = crClient.Get(context.Background(), nn, infraConfig)
	if err != nil {
		return err
	}
	fmt.Printf("using crclient: %v\n", infraConfig.Status.ControlPlaneTopology)
	fmt.Printf("using crclient: %v\n", infraConfig.Status.InfrastructureTopology)
```

And similarly with normal k8s constructs
```golang
	operatorConfigInformer := configinformer.NewSharedInformerFactoryWithOptions(configClient, 2*time.Second)
	infrastructureLister = operatorConfigInformer.Config().V1().Infrastructures().Lister()
	infraConfig, err := configClient.ConfigV1().Infrastructures().Get(context.Background(), "cluster", metav1.GetOptions{})
	if err != nil {
		return err
	}

	// fmt.Printf("%v\n", infraConfig)
	fmt.Printf("%v\n", infraConfig.Status.ControlPlaneTopology)
	fmt.Printf("%v\n", infraConfig.Status.InfrastructureTopology)
```

Any helper library would simply be a layer above what is already a seemingly simple API call. So I'm advocating we simply close this EP.

Reply:
@jmrodri I think your conclusion makes sense.

@varshaprasad96, what do you think?

Reply:
Yes, I agree. Closing this EP